crates/polars-lazy/src/physical_plan/expressions/ternary.rs

```diff
@@ -146,166 +121,168 @@ impl PhysicalExpr for TernaryExpr {
+                },
+                NotAggregated(_) => {
+                    let _ = ac.aggregated();
+                },
+                _ => {},
+            }
+        }
+
+        if has_non_unit_literal {
+            // Non-unit literals must be materialized per-group.
+            if state.verbose() {
+                println!("ternary agg: finish as iters due to non-unit literal")
+            }
+            return finish_as_iters(ac_truthy, ac_falsy, ac_mask);
+        }
+
+        let mut non_literal_acs = Vec::<&AggregationContext>::with_capacity(3);
+
+        for ac in [&ac_mask, &ac_truthy, &ac_falsy].into_iter() {
+            if !matches!(ac.agg_state(), Literal(_)) {
+                non_literal_acs.push(ac);
+            }
+        }
+
+        if non_literal_acs.is_empty() {
+            // All unit literals.
+            if state.verbose() {
+                println!(
+                    "ternary agg: finish all unit literals - expression could have been simplified"
+                )
+            }
+            ac_truthy.with_literal(
+                ac_truthy
+                    .series()
+                    .zip_with(ac_mask.series().bool()?, ac_falsy.series())?,
+            );
+
+            return Ok(ac_truthy);
+        }
 
-                if ac_falsy.is_literal() && self.falsy.as_expression().map(has_null) == Some(true) {
-                    let s = ac_truthy.aggregated();
-                    let ca = s.list().unwrap();
-                    check_length(ca, mask)?;
-                    let out = ca
-                        .into_iter()
-                        .zip(mask)
-                        .map(|(truthy, take)| if take? { truthy } else { None })
-                        .collect_trusted::<ListChunked>()
-                        .with_name(ac_truthy.series().name());
-                    ac_truthy.with_series(out.into_series(), true, Some(&self.expr))?;
-                    Ok(ac_truthy)
-                } else if ac_truthy.is_literal()
-                    && self.truthy.as_expression().map(has_null) == Some(true)
-                {
-                    let s = ac_falsy.aggregated();
-                    let ca = s.list().unwrap();
-                    check_length(ca, mask)?;
-                    let out = ca
-                        .into_iter()
-                        .zip(mask)
-                        .map(|(falsy, take)| if take? { None } else { falsy })
-                        .collect_trusted::<ListChunked>()
-                        .with_name(ac_truthy.series().name());
-                    ac_truthy.with_series(out.into_series(), true, Some(&self.expr))?;
-                    Ok(ac_truthy)
+        for (ac_l, ac_r) in non_literal_acs.iter().zip(non_literal_acs.iter().skip(1)) {
+            if std::mem::discriminant(ac_l.agg_state()) != std::mem::discriminant(ac_r.agg_state())
+            {
+                // Mix of AggregatedScalar and AggregatedList is done per group,
```

Comment:
Can't we also have `NonAggregated` here?

Reply:
Ah, I see.

Reply:
@nameexhaustion can you then add the trivial case, where all is non-aggregated, or all is non-aggregated + unit literal?


That will be much cheaper.